elementary/clients/teams/client.py

```diff
@@ -0,0 +1,110 @@
+from abc import ABC, abstractmethod
+from typing import Optional
+
+from pymsteams import cardsection, connectorcard, potentialaction  # type: ignore
+from retry import retry
+
+from elementary.config.config import Config
+from elementary.tracking.tracking_interface import Tracking
+from elementary.utils.log import get_logger
+
+logger = get_logger(__name__)
+
+OK_STATUS_CODE = 200
+
+
+class TeamsClient(ABC):
+    def __init__(
+        self,
+        tracking: Optional[Tracking] = None,
+    ):
+        self.client = self._initial_client()
+        self.tracking = tracking
+
+    @staticmethod
+    def create_client(
+        config: Config, tracking: Optional[Tracking] = None
+    ) -> Optional["TeamsClient"]:
+        if not config.has_teams:
+            return None
+        if config.teams_webhook:
+            return TeamsWebhookMessageBuilderClient(
+                webhook=config.teams_webhook, tracking=tracking
+            )
+        return None
+
+    @abstractmethod
+    def _initial_client(self):
+        raise NotImplementedError
+
+    def _initial_retry_handlers(self):
+        raise NotImplementedError
+
+    @abstractmethod
+    def send_message(self, **kwargs):
+        raise NotImplementedError
+
+    @abstractmethod
+    def title(self, title: str):
+        raise NotImplementedError
+
+    @abstractmethod
+    def text(self, text: str):
+        raise NotImplementedError
+
+    @abstractmethod
+    def addSection(self, section: cardsection):
+        raise NotImplementedError
+
+    @abstractmethod
+    def addPotentialAction(self, action: potentialaction):
+        raise NotImplementedError
+
+
+class TeamsWebhookClient(TeamsClient):
+    def __init__(
+        self,
+        webhook: str,
+        tracking: Optional[Tracking] = None,
+    ):
+        self.webhook = webhook
+        super().__init__(tracking)
+
+    def _initial_client(self):
+        return connectorcard(self.webhook)
+
+    @retry(tries=3, delay=1, backoff=2, max_delay=5)
+    def send_message(self, **kwargs) -> bool:
+        self.client.send()
+        response = self.client.last_http_response
+
+        if response.status_code == OK_STATUS_CODE:
+            return True
+        else:
+            logger.error(
+                "Could not post message to teams via webhook - %s. Error: %s",
+                {self.webhook},
+                {response.body},
+            )
+        return False
+
+
+class TeamsWebhookMessageBuilderClient(TeamsWebhookClient):
```

Comment:
The way `pymsteams` is setup is that there is no way to create a message **without** having a client connection. This does not allow for a similar setup as is done in Slack. This was the only way I could get it to work for now.

The ideal solution would have a TeamsMessageBuilder class that directly modifies the payload, a good starting place is the [pymsteams code](https://github.com/rveachkc/pymsteams/blob/master/pymsteams/__init__.py). 

Since this was a MVP I still think this works good enough.

Reply:
Yeah I understand, still don't you think it makes more sense that way:
```
class TeamsWebhookMessageBuilder(ABC):
    def __init__(client: TeamsWebhookClient):
        self.client = client
````
?

Reply:
Aah now I understand what you propose, yeah I think that it makes more sense to do it in that way. Think that it is more neat. Will do that, but probably tomorrow

Reply:
I've been struggling with it for some time now. And now I'm starting to doubt if it makes sense to have a seperate class to create the message on at all. 

Now I think about it for some time, I would like to suggest to only create a class `TeamsClient`, and this class does handle the connection to teams and sending the message, as well as creating the json `payload`. This is completely in line with `pymsteams` and leads to the least amount of redundant code.

Do you agree with this?

Reply:
Hey @FrankTub, chiming in here 🙂 
I saw this thread only now after I commented on that part as well.
I see the problem you are raising. Another approach could be to create a `TeamsMessageBuilder` class that is getting the client / connectcard as a param and return a connectcard.
That way you can build everything in the message builder scope, and send the connectcard using the client.
I know this could sound a bit overkill, but the idea of the message builder is that you can create different ones for the same integration and control the messages templates behaviours differently and easily.

In case this is not working well and problematic to do, I think that at least we should move the `title / text / etc` methods into the `TeamsWebhookClient` class and remove them from the abstract `TeamsClient` class as they might not be relevant for the other clients (if the will we will add them back to the abstract one) 🙂

Reply:
@IDoneShaveIt regarding:

> I see the problem you are raising. Another approach could be to create a TeamsMessageBuilder class that is getting the client / connectcard as a param and return a connectcard.

I think this leads to a lot of duplicated code. Since you need to add the functions of `pymsteams` in the `TeamsClient` class, as well as in the `TeamsMessageBuilder` class. And basicly it is all just copy pasting code of what `pymsteams` already does for you.

Regarding:

> In case this is not working well and problematic to do, I think that at least we should move the title / text / etc methods into the TeamsWebhookClient class and remove them from the abstract TeamsClient class as they might not be relevant for the other clients (if the will we will add them back to the abstract one)

I implemented it in that way initially, however `mypy` does not approve.

## Proposal

The way I see it we could do two things and would like your opinion what you think is best to do.

1) Remove all `TeamsClient` classes and just use `pymsteams` where the `TeamsClient` was used up till this point. This leads to the least amount of redundant code, but is not in line with how the rest of the code base.
2) Just have a `TeamsWebhookMessageBuilderClient` class without a `TeamsClient`. This is just a wrapper around `pymsteams` and makes it more similar to the way the Slack client is working.

It would be something like:

```py
class TeamsWebhookMessageBuilderClient(ABC):
    def __init__(
        self,
        webhook: str,
        tracking: Optional[Tracking] = None,
    ):
        self.webhook = webhook
        self.tracking = tracking
        self.client = self._initial_client()

    @staticmethod
    def create_client(
        config: Config, tracking: Optional[Tracking] = None
    ) -> Optional["TeamsWebhookMessageBuilderClient"]:
        if not config.has_teams:
            return None
        if config.teams_webhook:
            return TeamsWebhookMessageBuilderClient(
                webhook=config.teams_webhook, tracking=tracking
            )
        return None

    @retry(tries=3, delay=1, backoff=2, max_delay=5)
    def send_message(self, **kwargs) -> bool:
        self.client.send()
        response = self.client.last_http_response

        if response.status_code == OK_STATUS_CODE:
            return True
        else:
            logger.error(
                "Could not post message to teams via webhook - %s. Error: %s",
                {self.webhook},
                {response.body},
            )
        return False

    def _initial_client(self):
        return connectorcard(self.webhook)

    def title(self, title: str):
        self.client.title(title)

    def text(self, text: str):
        self.client.text(text)

    def addSection(self, section: cardsection):
        self.client.addSection(section)

    def addPotentialAction(self, action: potentialaction):
        self.client.addPotentialAction(action)
```

Would like to know your view on this! 🙏

Reply:
> I think this leads to a lot of duplicated code. Since you need to add the functions of `pymsteams` in the `TeamsClient` class, as well as in the `TeamsMessageBuilder` class. And basicly it is all just copy pasting code of what `pymsteams` already does for you.

You don't have to create `title / text / etc` at the `TeamsClient`. Because you pass the `TeamsClient` to the message builder (or a connectcard) you can just do `client/connectcard.text` for example.
That way you can create all the `title / text / etc` methods only at the message builder.

> Remove all TeamsClient classes and just use pymsteams where the TeamsClient was used up till this point. This leads to the least amount of redundant code, but is not in line with how the rest of the code base.

The reason I do want to have a `TeamsClient` is that in case teams changes their sdk it will contain all the changes we need to do in one place (instead across the platform). Even if it just a wrapper for `pymsteams`, it makes it more maintainable 🙂 

  > Just have a TeamsWebhookMessageBuilderClient class without a TeamsClient. This is just a wrapper around pymsteams and makes it more similar to the way the Slack client is working.

In that case I prefer having `TeamsClient` and `TeamsWebhookClient` and skip the `TeamsWebhookMessageBuilder`. It will look the same as the current implementation, but with a different name 🙂

If it sounds valid to you, I think that passing the connectcard to the message builder is better, but we can go with the second suggestion as well and skip the  `TeamsWebhookMessageBuilder` for now 🙂

Reply:
@IDoneShaveIt I got it working as per your preferred way. Created a seperate `TeamsAlertMessageBuilder`. The thing I don't like about it is that in the client `TeamsClient` and `TeamsWebhookClient` it is required to define the methods to modify your message as well. That being said I think it makes more sense this way.

Curious to hear what you think about it!